include/swift/SIL/Mangle.h

```diff
@@ -104,6 +104,8 @@ template <typename SubType>
 class SpecializationMangler : public SpecializationManglerBase {
   SubType *asImpl() { return static_cast<SubType *>(this); }
 public:
+  CanSILFunctionType CanSILFnTy;
```

Comment:
Yes, it could be done with just a GenericSignature. Though currently, for full specializations we do not generate a generic signature, because it is not required. To use a GenericSignature-based approach we'd need to generate new generic signatures always, even in cases of full specializations.